Update upstreaming scripts

Prof-Bloodstone opened the pull request:
Script cleanup and fixes. Notable behavioral changes:
1. I've removed a "feature" where running `upstreamMerge.sh` without checked out submodules deleted your workspace.
2. Added a way to disable patch rebuild process - useful for automation ("Aikar script").

I've tried my best to not use too modern Bash features, everything should be very standard.

CC for script review: @Proximyst

Prof-Bloodstone committed:
Update upstreamMerge script

No longer nukes workspace if submodules are not checked out

Prof-Bloodstone committed:
Update upstreamCommit script

Proximyst commented on scripts/upstreamCommit.sh:
```diff
@@ -1,38 +1,41 @@
-) || exit 1
+  if [ -n "${updated:-}" ]; then
+    log="${UP_LOG_PREFIX:-}Updated Upstream ($updated)\n\n${disclaimer}${logsuffix}"
+    echo -e "${log}" | git commit -F -
```
`echo -e` is not available on whatever macOS uses I believe

Prof-Bloodstone replied:
I removed it, but then added back because it was used before anyway. Can replace with `printf` builtin.

Proximyst replied:
It's kind of up to cat to decide, honestly. I don't have a mac to test.

Prof-Bloodstone replied:
Replaced with printf builtin. Using `%b` format specifier (available in Bash, Ksh and BB) that will make sure to only expand escaped characters and not try to parse printf format specifiers.
See: <https://github.com/koalaman/shellcheck/wiki/SC2059>

Prof-Bloodstone committed:
printf > echo

https://unix.stackexchange.com/questions/65803/why-is-printf-better-than-echo

Proximyst reviewed (approved):
![](https://media3.giphy.com/media/13zeE9qQNC5IKk/giphy.webp)

kennytv reviewed (approved):
Pufferfish Seal of Approval for Windows 10 and WSL2 with some Ubuntu version
![pufferfish](https://user-images.githubusercontent.com/28825609/125091424-ad491780-e0d0-11eb-8607-16ba08985da8.png)

electronicboy commented on scripts/upstreamMerge.sh:
```diff
@@ -1,41 +1,44 @@
+  update Bukkit
+  update CraftBukkit
+  update Spigot
+  if [[ "${update_all}" == @(all|a|yes|y) ]]; then
```
```
 15:21:29  shane@Shanes-iMac  ...work/paper/Paper   update_scripts ✔ 
$ ./scripts/upstreamMerge.sh
./scripts/upstreamMerge.sh: line 29: syntax error in conditional expression: unexpected token `('
```

Prof-Bloodstone replied:
Fixed this. Tested with bash 3.2.57 and found a bug with `command` builtin, so I also removed that part of the code. Seems to work fine now.

Prof-Bloodstone committed:
Make upstream script work on bash 3.2.57

Support the poor souls stuck using it

Apparently there's a bug with 'command' immediately exiting whole script on error...

Prof-Bloodstone committed:
Actually disable patch filtering on upstream commits

electronicboy commented on scripts/upstreamMerge.sh:
```diff
@@ -37,7 +37,7 @@ function main {
     ./gradlew cleanCache || exit 1 # todo: Figure out why this is necessary
     ./gradlew applyPatches -Dpaperweight.debug=true || exit 1
-    ./gradlew rebuildPatches || exit 1
+    ./gradlew rebuildPatches --filter-patches=false || exit 1
```
```
See "git help gc" for manual housekeeping.
See "git help gc" for manual housekeeping.
   Patches applied cleanly to Paper-Server

BUILD SUCCESSFUL in 9m 34s
36 actionable tasks: 31 executed, 4 from cache, 1 up-to-date

FAILURE: Build failed with an exception.

* What went wrong:
Problem configuring task :rebuildPatches from command line.
> Unknown command-line option '--filter-patches'.

* Try:
Run gradlew help --task :rebuildPatches to get task usage details. Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
```
apparently this is not a thing?

MiniDigger replied:
cool cool

![https://i.imgur.com/P903yvW.png](https://i.imgur.com/P903yvW.png)

https://github.com/PaperMC/paperweight/blob/f5d09e8d1d162a8371dcae79791abe3c12ff3013/paperweight-lib/src/main/kotlin/tasks/RebuildPaperPatches.kt#L52

gradle moment?

MiniDigger replied:
ok, turns out you can't pass that option to the rebuildPatches task, you need to pass it to the rebuildApiPatches and rebuildServerPatches tasks, this is a documented limitation of gradle options https://docs.gradle.org/current/userguide/custom_tasks.html#limitations

so I guess we have to do this
```suggestion
    ./gradlew rebuildApiPatches --filter-patches=false || exit 1
    ./gradlew rebuildServerPatches --filter-patches=false || exit 1
```

Prof-Bloodstone replied:
I have no idea how I tested it to not catch it. Also, apparently, you can't disable any gradle command-line options through CLI so this doesn't work at all: https://github.com/gradle/gradle/issues/4311

Prof-Bloodstone replied:
Reverted this for now, as this was additional thing I did and don't want it to hold other PR and internal stuff blocked.

Prof-Bloodstone replied:
Updated to use the property introduced in latest paperweight: https://github.com/PaperMC/paperweight/commit/1e3d63a29b659cde2ca8ee4944cbfc888b821541
It does nothing for now, but will automagically work when Paper updates the paperweight version it uses.
Also, at @jpenilla request, the script will rebuild patches without filtering before updating submodules, then do it again once submodules are updated. If submodules introduce no changes, changes to all patches will be reverted.

I could probably also make it first check, if any of the submodules requested for update have changes (do `git fetch` and `git diff --exit-code --quiet origin/master`), and only if changes are expected, do the whole process?

Prof-Bloodstone committed:
Revert disabling of patch filtering as it doesn't work

Prof-Bloodstone committed:
Revert "Revert disabling of patch filtering as it doesn't work"

This reverts commit 150374501b72af4eebca907b6bed5bad42bfe8f4.

Prof-Bloodstone committed:
Use future Property to skip filtering patches and rebuild before update

Prof-Bloodstone committed:
Don't pre-rebuild patches if there will be no changes

Prof-Bloodstone committed:
Remove debug 'set -x'

Prof-Bloodstone committed:
Add function to update multiple submodules (see #upstream-updates)

electronicboy commented on scripts/upstreamMerge.sh:
```diff
@@ -1,41 +1,98 @@
+      echo "Rebuilding patches without filtering to improve apply ability"
+      cd "${basedir}"
+      ./gradlew cleanCache || return $? # todo: Figure out why this is necessary
+      apply_rebuild || return $?
```
2nd rebuild shouldn't be filtering the patches, otherwise it induces much noise in the final patches
